core/src/main/scala/besom/internal/ResourceOps.scala

```diff
@@ -37,34 +52,30 @@ class ResourceOps(using ctx: Context, mdc: BesomMDC[Label]):
     options: ResourceOptions,
     remote: Boolean
   ): Result[R] =
-    summon[ResourceDecoder[R]].makeResolver.flatMap { (resource, resolver) =>
-      // this order is then repeated in registerReadOrGetResource
-      val modeResult = options.hasURN.zip(options.hasImportId).map { case (hasUrn, hasImportId) =>
-        if hasUrn then "Getting" else if hasImportId then "Reading" else "Registering"
-      }
-
-      modeResult.flatMap { mode =>
-        log.debug(s"$mode resource, trying to add to cache...") *>
-          ctx.resources.cacheResource(typ, name, args, options, resource).flatMap { addedToCache =>
-            if addedToCache then
-              for
-                options <- options.resolve
-                _       <- log.debug(s"$mode resource, added to cache...")
-                state   <- createResourceState(typ, name, resource, options, remote)
-                _       <- log.debug(s"Created resource state")
-                _       <- ctx.resources.add(resource, state)
-                _       <- log.debug(s"Added resource to resources")
-                inputs  <- prepareResourceInputs(resource, state, args, options)
-                _       <- log.debug(s"Prepared inputs for resource")
-                _       <- addChildToParentResource(resource, options.parent)
-                _       <- log.debug(s"Added child to parent (isDefined: ${options.parent.isDefined})")
-                _       <- registerReadOrGetResource(resource, state, resolver, inputs, options, remote)
-              yield resource
-            else ctx.resources.getCachedResource(typ, name, args, options).map(_.asInstanceOf[R])
-          }
-      }
+    resolveMode(options).flatMap { mode =>
+      def runSideEffects =
+        for
+          (resource, resolver) <- ResourceDecoder.forResource[R].makeResourceAndResolver
+          _                    <- log.debug(s"$mode resource ${mode.suffix}")
+          options              <- options.resolve
+          _                    <- log.debug(s"$mode resource, added to cache...")
+          state                <- createResourceState(typ, name, resource, options, remote)
+          _                    <- log.debug(s"Created resource state")
+          _                    <- ctx.resources.add(resource, state)
+          _                    <- log.debug(s"Added resource to resources")
+          inputs               <- prepareResourceInputs(resource, state, args, options)
+          _                    <- log.debug(s"Prepared inputs for resource")
+          _                    <- addChildToParentResource(resource, options.parent)
+          _                    <- log.debug(s"Added child to parent (isDefined: ${options.parent.isDefined})")
+          _                    <- registerReadOrGetResource(resource, state, resolver, inputs, options, remote)
+        yield resource
+
+      mode match
```

Comment:
tbh the only reason is that get is the only one that doesn't crash on repeated evaluations and I want to memoize as little as possible (only where calls are not idempotent and crash deployments) because it _might_ (not necessarily true) be possible to get two different values from two gets executed in different phases of a deployment.

Reply:
I assumed read would have the same property but no, it doesn't, multiple reads of the same resource crash.